packages/SwingSet/src/vats/timer/vat-timer.js

```diff
@@ -1,98 +1,854 @@
+  // return the smallest value of `start + N * interval` after now
+  if (now < start) {
+    return start;
+  }
+  return now + interval - ((now - start) % interval);
+}
+
+export function buildRootObject(vatPowers, _vatParameters, baggage) {
   const { D } = vatPowers;
-  const repeaters = new Map();
 
-  async function createTimerService(timerNode) {
-    /** @type {TimerService} */
-    const timerService = Far('timerService', {
-      getCurrentTimestamp() {
-        return Nat(D(timerNode).getLastPolled());
-      },
-      setWakeup(baseTime, handler) {
-        baseTime = TimeMath.toAbs(baseTime);
-        assert(passStyleOf(handler) === 'remotable', 'bad setWakeup() handler');
-        return D(timerNode).setWakeup(baseTime, handler);
-      },
-      // can be used after setWakeup(h) or schedule(h)
-      removeWakeup(handler) {
+  // we use baggage to retain a device reference across upgrades
+  let timerDevice;
+  if (baggage.has('timerDevice')) {
+    timerDevice = baggage.get('timerDevice');
+  }
+  function insistDevice() {
+    assert(timerDevice, 'TimerService used before createTimerService()');
+  }
+
+  // first, create the KindHandles for all our durable objects
+  const timerServiceHandle = provideKindHandle(baggage, 'timerServiceHandle');
+  const oneShotEventHandle = provideKindHandle(baggage, 'oneShotEventHandle');
+  const promiseEventHandle = provideKindHandle(baggage, 'promiseEventHandle');
+  const repeaterEventHandle = provideKindHandle(baggage, 'repeaterEventHandle');
+  const repeaterHandle = provideKindHandle(baggage, 'repeaterHandle');
+  const notifierHandle = provideKindHandle(baggage, 'notifierHandle');
+
+  // then the two durable tables
+
+  // The 'schedule' maps upcoming timestamps to the Event that should
+  // be fired. We rely upon the earlier-vs-later sortability of BigInt
+  // keys, and upon our Stores performing efficient iteration.
+
+  /** @type {Schedule} */
+  const schedule = provide(baggage, 'schedule', () =>
+    makeScalarBigMapStore('schedule', { durable: true }),
+  );
+
+  // 'cancels' maps cancel handles to Events that will be
+  // removed. Each event knows its own .scheduled time, so we can find
+  // and remove it from 'schedule'
+
+  /** @type {CancelTable} */
+  const cancels = provide(baggage, 'cancels', () =>
+    makeScalarBigWeakMapStore('cancels', { durable: true }),
+  );
+
+  // Then an *ephemeral* WeakMap to hang on to the ephemeral Promise
+  // resolve/reject functions for delay/wakeAt. We can't hold these
+  // bare functions in the (durable) PromiseEvent, but we *can* use
+  // the PromiseEvent as a key to fetch them when the event
+  // fires. It's ok for these to be ephemeral: all promises get
+  // rejected (with { name: 'vatUpgraded' }) during an upgrade, so if
+  // the timer fires *after* an upgrade, we no longer need to reject
+  // it ourselves.
+
+  /** @type {WakeupPromiseTable} */
+  const wakeupPromises = makeLegacyWeakMap('promises');
```

Comment:
Why not
```suggestion
  const wakeupPromises = makeScalarWeakMapStore('promises');
```
?

It seems the reason is that the values are PromiseControls, and PromiseControls are not declared `Far`. I'd rather have them declared `Far` than use a legacyMapStore.

Reply:
I see it's done. Resolving.